File: gandalf/runner.py
"""Sandboxing and grading of a single exercise.

Every exercise is copied (permissions and symlinks preserved) into a temporary
directory before any check runs, so a test can extract a tarball or delete a
file without ever touching the student's work.
"""

import importlib.util
import os
import shutil
import stat
import subprocess
import sys
import tempfile
import time


class Check(object):
    def __init__(self, label, ok, detail=""):
        self.label = label
        self.ok = ok
        self.detail = detail


class Grader(object):
    """Handed to every exercise module's `run()`."""

    def __init__(self, sandbox, origin):
        self.sandbox = sandbox    # temp copy, safe to mutate
        self.origin = origin      # real exercise dir, read only
        self.checks = []
        self.notes = []

    # --- assertions ------------------------------------------------------
    def check(self, label, ok, detail=""):
        self.checks.append(Check(label, bool(ok), detail))
        return bool(ok)

    def note(self, message):
        """Advisory line, not worth any point."""
        self.notes.append(message)

    # --- filesystem helpers ---------------------------------------------
    def path(self, *parts):
        return os.path.join(self.sandbox, *parts)

    def exists(self, name):
        return os.path.lexists(self.path(name))

    def is_file(self, name):
        return os.path.isfile(self.path(name))

    def size(self, name):
        try:
            return os.stat(self.path(name)).st_size
        except OSError:
            return None

    def mode(self, name):
        try:
            return stat.filemode(os.lstat(self.path(name)).st_mode)
        except OSError:
            return None

    def read_bytes(self, name):
        try:
            with open(self.path(name), "rb") as fh:
                return fh.read()
        except (OSError, IOError):
            return None

    def sh(self, command, timeout=10):
        """Run a shell command inside the sandbox. Returns (rc, stdout, stderr)."""
        try:
            proc = subprocess.run(
                command, shell=True, cwd=self.sandbox, timeout=timeout,
                stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            )
            return proc.returncode, proc.stdout, proc.stderr
        except subprocess.TimeoutExpired:
            return 124, b"", b"timeout"

    # --- scoring ---------------------------------------------------------
    @property
    def passed(self):
        return sum(1 for c in self.checks if c.ok)

    @property
    def total(self):
        return len(self.checks)

    @property
    def failures(self):
        return [c for c in self.checks if not c.ok]


def _force_writable(func, path, _exc):
    try:
        os.chmod(os.path.dirname(path), 0o700)
        os.chmod(path, 0o700)
        func(path)
    except OSError:
        pass


def load_module(exercise):
    spec = importlib.util.spec_from_file_location(
        "gandalf_tests_%s_%s" % (exercise.quest.name, exercise.dir),
        exercise.module_path,
    )
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def grade(exercise, exercise_path):
    """Run one exercise. Returns (grader, runtime_seconds)."""
    sandbox_root = tempfile.mkdtemp(prefix="gandalf-")
    sandbox = os.path.join(sandbox_root, exercise.dir)
    try:
        shutil.copytree(exercise_path, sandbox, symlinks=True)
        os.chmod(sandbox, 0o755)

        module = load_module(exercise)
        grader = Grader(sandbox, exercise_path)

        start = time.perf_counter()
        module.run(grader)
        runtime = time.perf_counter() - start
        return grader, runtime
    finally:
        _rmtree(sandbox_root)


def _rmtree(path):
    if sys.version_info >= (3, 12):
        shutil.rmtree(path, onexc=_force_writable)
    else:
        shutil.rmtree(path, onerror=_force_writable)

File: lib/runner.js
"use strict";

const fs = require("fs");
const os = require("os");
const path = require("path");
const vm = require("vm");
const { spawnSync } = require("child_process");
const { filemode } = require("./filemode");

class Check {
  constructor(label, ok, detail) {
    this.label = label;
    this.ok = !!ok;
    this.detail = detail || "";
  }
}

class Grader {
  constructor(sandbox, origin) {
    this.sandbox = sandbox; // temp copy, safe to mutate
    this.origin = origin; // real exercise dir, read only
    this.checks = [];
    this.notes = [];
    this.testCases = [];
  }

  // --- assertions --------------------------------------------------------
  check(label, ok, detail) {
    ok = !!ok;
    this.checks.push(new Check(label, ok, detail));
    return ok;
  }

  note(message) {
    this.notes.push(message);
  }

  /**
   * Records one test case for the "Checks detail Report" table (used by
   * exercises graded like Qwasar's my_levenshtein_web: several labeled
   * input/output/return-value comparisons per exercise, each shown as its
   * own sub-table). Also feeds the normal pass/fail tally via `check()`, so
   * scoring stays consistent whether or not detail is displayed.
   *
   * `record`: { label, input: [...], expectedOutput, expectedReturn, output,
   *             returnValue, passed, detail }
   */
  testCase(record) {
    this.testCases.push(record);
    return this.check(record.label || `test case ${this.testCases.length - 1}`, record.passed, record.detail);
  }

  // --- filesystem helpers --------------------------------------------------
  path(...parts) {
    return path.join(this.sandbox, ...parts);
  }

  exists(name) {
    try {
      fs.lstatSync(this.path(name));
      return true;
    } catch {
      return false;
    }
  }

  isFile(name) {
    try {
      return fs.statSync(this.path(name)).isFile();
    } catch {
      return false;
    }
  }

  size(name) {
    try {
      return fs.statSync(this.path(name)).size;
    } catch {
      return null;
    }
  }

  /** Permission string like "-r--r-xr-x", following symlinks like `ls -l`. */
  mode(name) {
    try {
      return filemode(fs.statSync(this.path(name)).mode);
    } catch {
      return null;
    }
  }

  readBytes(name) {
    try {
      return fs.readFileSync(this.path(name));
    } catch {
      return null;
    }
  }

  readText(name, encoding = "utf8") {
    const buf = this.readBytes(name);
    return buf === null ? null : buf.toString(encoding);
  }

  sh(command, { timeout = 10000 } = {}) {
    const result = spawnSync("/bin/sh", ["-c", command], {
      cwd: this.sandbox,
      timeout,
    });
    if (result.error && result.error.code === "ETIMEDOUT") {
      return { code: 124, stdout: Buffer.alloc(0), stderr: Buffer.from("timeout") };
    }
    return {
      code: result.status === null ? 1 : result.status,
      stdout: result.stdout || Buffer.alloc(0),
      stderr: result.stderr || Buffer.alloc(0),
    };
  }

  // --- scoring -------------------------------------------------------------
  get passed() {
    return this.checks.filter((c) => c.ok).length;
  }

  get total() {
    return this.checks.length;
  }

  get failures() {
    return this.checks.filter((c) => !c.ok);
  }
}

function copyTree(src, dest) {
  const st = fs.lstatSync(src);
  if (st.isSymbolicLink()) {
    fs.symlinkSync(fs.readlinkSync(src), dest);
  } else if (st.isDirectory()) {
    fs.mkdirSync(dest, { mode: st.mode & 0o777 });
    for (const entry of fs.readdirSync(src)) {
      copyTree(path.join(src, entry), path.join(dest, entry));
    }
    fs.chmodSync(dest, st.mode & 0o777);
  } else if (st.isFile()) {
    fs.copyFileSync(src, dest);
    fs.chmodSync(dest, st.mode & 0o777);
  }
}

function rmForce(target) {
  fs.rmSync(target, { recursive: true, force: true, maxRetries: 3 });
}

function loadModule(exercise) {
  // Bypass require's cache so re-running the same exercise in one process
  // (tests, `--list` follow-ups, etc.) always re-reads the file from disk.
  delete require.cache[require.resolve(exercise.modulePath)];
  return require(exercise.modulePath);
}

/** Runs one exercise against a sandboxed copy. Returns { grader, runtime }. */
function grade(exercise, exercisePath) {
  const sandboxRoot = fs.mkdtempSync(path.join(os.tmpdir(), "gandalf-"));
  let sandbox;
  try {
    if (exercise.dir === ".") {
      // exercisePath IS the exercise root (no exNN subfolder) -- copy its
      // contents directly into sandboxRoot, which mkdtempSync already made.
      sandbox = sandboxRoot;
      for (const entry of fs.readdirSync(exercisePath)) {
        copyTree(path.join(exercisePath, entry), path.join(sandbox, entry));
      }
    } else {
      sandbox = path.join(sandboxRoot, exercise.dir);
      copyTree(exercisePath, sandbox);
    }

    const mod = loadModule(exercise);
    const grader = new Grader(sandbox, exercisePath);

    const start = process.hrtime.bigint();
    mod.run(grader);
    const runtime = Number(process.hrtime.bigint() - start) / 1e9;
    return { grader, runtime };
  } finally {
    rmForce(sandboxRoot);
  }
}

module.exports = { Grader, Check, grade, copyTree, vm };
